### PR TITLE
Dev/saars/container id fallback

### PR DESCRIPTION
**Pull request by xiaomi7732**
Address #246

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+# XML config files
+[*.{props,targets,ruleset,config,nuspec,resx,vsixmanifest,vsct}]
+indent_size = 2
+
+# Code files
+[*.{cs,csx,vb,vbx}]
+indent_size = 4
+insert_final_newline = true
+charset = utf-8-bom
```
Comment (karolz-ms) [inline .editorconfig:21]: Consider just `utf8`. BOM is not recommended for `utf8` encoding

File: src/ApplicationInsights.Kubernetes/ContainerIdProviders/NullContainerIdProvider.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿#nullable enable
+
+namespace Microsoft.ApplicationInsights.Kubernetes.ContainerIdProviders
+{
+    internal class EmptyContainerIdProvider : IContainerIdProvider
```
Comment (karolz-ms) [inline src/ApplicationInsights.Kubernetes/ContainerIdProviders/NullContainerIdProvider.cs:5]: nit: class name does not correspond to file name

**Review by karolz-ms — APPROVED** [review]
Looks good, just minor change suggestions

File: .editorconfig (hunk 1 above)
Comment (xiaomi7732) [reply .editorconfig:21]: Yes, no one likes bom, and I will dig out who set it. :-D

File: src/ApplicationInsights.Kubernetes/ContainerIdProviders/NullContainerIdProvider.cs (hunk 2 above)
Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/ContainerIdProviders/NullContainerIdProvider.cs:5]: > Consider naming the class `WindowsContainerIdProvider`. The fact that it is currently unable to retrieve a container ID may change in future.

Thanks for notice the mismatch.

I renamed the file to `EmptyContainerIdProvider`. If/When there's new way to retrieve the container id for windows container, that would be the new class of `WindowsContaienrIdProvider` being appended.

Currently, the name of `EmptyContainerIdProvider` reveals what the class does.

File: src/ApplicationInsights.Kubernetes/EntityExtensions/PodExtensions.cs [hunk 3]
```diff
@@ -8,25 +10,23 @@
...
     internal static class PodExtensions
     {
-        public static ContainerStatus GetContainerStatus(this K8sPod self, string containerId)
-        {
-            ContainerStatus result = self.Status.ContainerStatuses?.FirstOrDefault(
-                cs => !string.IsNullOrEmpty(cs.ContainerID) && cs.ContainerID.EndsWith(containerId, StringComparison.Ordinal));
-            return result;
-        }
+        public static ContainerStatus? GetContainerStatus(this K8sPod self, string containerId)
+            => containerId == string.Empty ? null // Special case when container id is an empty string,
```
[the remark this answers is not in the document]
Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/EntityExtensions/PodExtensions.cs]: Sure. Still try to get a feeling about null annotation. :-)

File: src/ApplicationInsights.Kubernetes/Extensions/KubernetesServiceCollectionBuilder.cs [hunk 4]
```diff
@@ -75,10 +77,14 @@ protected virtual void RegisterSettingsProvider(IServiceCollection serviceCollec
         {
             if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
             {
+                serviceCollection.TryAddEnumerable(ServiceDescriptor.Singleton<IContainerIdProvider, CGroupContainerIdProvider>());
```
[the remark this answers is not in the document]
Comment (xiaomi7732) [reply src/ApplicationInsights.Kubernetes/Extensions/KubernetesServiceCollectionBuilder.cs:82]: Good point.

**Pull request merged**